File: docs/js/bst/bst.js
/**
 * Created by wxy on 17-9-16.
 */
class Node {
    constructor(k, v, parent, level) {
        this.k = k;
        this.v = v;

        this.parent = parent;
        this.level = level;

        this.left = null;
        this.right = null;
    }

}
class BST {
    constructor() {
        this.root = null;
        this.count = 0;
    }


    nodeLevel(k) {
        var node = this.search(k)
        return this.__nodeLevel(node)
    }

    __nodeLevel(node) {
        if (!node)
            return -1;
        var l = 0;
        while (node.parent) {
            l++;
            node = node.parent;
        }
        return l
    }

    insert(k, v) {
        this.root = this.__insert(this.root, k, v)
    }

    __insert(node, k, v, parent = null) {
        if (!node) {
            this.count++;
            var parentLevel = this.__nodeLevel(parent);
            return new Node(k, v, parent, parentLevel + 1)
        }
        if (k === node.k)
            node.v = v;
        if (k < node.k)
            node.left = this.__insert(node.left, k, v, node);
        if (k > node.k)
            node.right = this.__insert(node.right, k, v, node);
        return node
    }

    search(k) {
        return this.__search(this.root, k);
    }

    __search(node, k) {
        if (!node)
            return null;
        if (k === node.k)
            return node;
        else if (k < node.k)
            return this.__search(node.left, k);
        else
            return this.__search(node.right, k);
    }

    preOrder() {
        this.__preOrder(this.root);
    }

    __preOrder(node) {
        if (!node)
            return;
        log(node.k, node.v, node.level);
        this.__preOrder(node.left);
        this.__preOrder(node.right);
    }

    inOrder() {
        this.__inOrder(this.root);
    }

    __inOrder(node) {
        if (!node)
            return;
        log(node.k, node.v, node.level);
        this.__inOrder(node.left);
        log(node.k, node.v, node.level);
        this.__inOrder(node.right);
    }

    levelOrder() {
        var l = [];
        l.push(this.root);
        while (l.length > 0) {
            var node = l.shift();
            log(node.k, node.v, node.level);
            if (node.left) l.push(node.left);
            if (node.right) l.push(node.right);
        }
    }

    minimum() {
        return this.__minimum(this.root).k
    }

    __minimum(node) {
        while (node.left)
            node = node.left;
        return node
    }

    maximum() {
        if (this.count) {
            var node = this.root
            while (node.right)
                node = node.right;
            return node
        }
    }

    removeMin() {
        this.root = this.__removeMin(this.root)
    }

    __removeMin(node) {
        // debugger
        if (!node)
            return
        if (!node.left) {
            this.count--
            // 同时修改parent
            if (node.right)
                node.right.parent = node.parent;
            return node.right

        }
        node.left = this.__removeMin(node.left)
        return node
    }

    removeMax() {
        this.root = this.__removeMax(this.root)
    }

    __removeMax(node) {
        if (!node)
            return
        if (!node.right) {
            this.count--
            // 同时修改parent
            if (node.left)
                node.left.parent = node.parent;
            return node.left

        }
        node.right = this.__removeMax(node.right)
        return node
    }

    remove(k) {
        if (!this.count)
            return;
        this.root = this.__remove(this.root, k)
        if (this.root)
            this.root.level = 0
    }

    __remove(node, k) {
        if (!node)
            return;
        if (k < node.k) {
            node.left = this.__remove(node.left, k);
            return node
        }
        if (k > node.k) {
            node.right = this.__remove(node.right, k);
            return node
        }

        if (k === node.k) {
            if (!node.left) {
                this.count--;
                if (node.right)
                    node.right.parent = node.parent;
                return node.right
            }
            else if (!node.right) {
                this.count--;
                if (node.left)
                    node.left.parent = node.parent;
                return node.left
            }
            else {//左右都有节点
                // debugger
                // successor取代node，父亲和孩子都要修改
                var successor = this.__minimum(node.right)

                successor.right = this.__removeMin(node.right)
                successor.left = node.left

                successor.parent = node.parent;

                if (successor.left)
                    successor.left.parent = successor;
                if (successor.right)
                    successor.right.parent = successor;
                successor.level = this.__nodeLevel(successor)

                return successor
            }
        }

    }


}

File: practice_python/BST/BST.py
from random import randint


class Node(object):
    def __init__(self, k, left=None, right=None):
        self.k = k
        self.left = left
        self.right = right

    def __repr__(self):
        return '{},(l:{},r:{})'.format(self.k, self.left, self.right)
        # left = 'l:' + str(self.left.k) if self.left else ''
        # right = 'r:' + str(self.right.k) if self.right else ''
        # return '{},({},{})'.format(self.k, left, right)


def do_something(node):
    print node


class BST(object):
    def __init__(self):
        self.root = None
        self.count = 0

    def insert(self, k):
        self.root = self.__insert(self.root, k)

    def __insert(self, node, k):
        if not node:
            self.count += 1
            return Node(k)
        if k == node.k:
            return node
        elif k < node.k:
            node.left = self.__insert(node.left, k)
        else:
            node.right = self.__insert(node.right, k)
        return node

    def pre_order(self):
        self.__pre_order(self.root)

    def __pre_order(self, node):
        if not node:
            return
        do_something(node)
        if node.left:
            self.__pre_order(node.left)
        if node.right:
            self.__pre_order(node.right)

    def post_order(self):
        self.__post_order(self.root)

    def __post_order(self, node):
        if not node:
            return
        if node.left:
            self.__post_order(node.left)
        if node.right:
            self.__post_order(node.right)
        do_something(node)

    def in_order(self):
        print 'inOrder'
        self.__in_order(self.root)

    def __in_order(self, node):
        if not node:
            return
        if node.left:
            self.__in_order(node.left)
        do_something(node)
        if node.right:
            self.__in_order(node.right)

    def level_order(self):
        q = []
        node = self.root
        q.append(node)
        while q:
            node = q.pop(0)
            do_something(node)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

    def minimum(self):
        if not self.root:
            return
        return self.__minimum(self.root)

    def __minimum(self, node):
        if not node.left:
            return node
        else:
            return self.__minimum(node.left)

    def remove_min(self):
        self.root = self.__remove_min(self.root)

    def __remove_min(self,node):
        if not node:
            return
        if not node.left:
            self.count-=1
            return node.right

        node.left = self.__remove_min(node.left)
        return node
    def contain(self, k):
        node = self.root
        while node:
            if k == node.k:
                return True
            elif k < node.k:
                node = node.left
            else:
                node = node.right
        return False

    def remove(self, k):
        if self.contain(k):
            self.root = self.__remove(self.root, k)

    def __remove(self, node, k):
        if not node:
            return
        if k < node.k:
            node.left = self.__remove(node.left,k)
            return node
        elif k > node.k:
            node.right = self.__remove(node.right,k)
            return node
        else:
            if not node.left:
                 return node.right
            elif not node.right:
                return node.left
            else:
                print 'test'
                successor = Node(self.__minimum(node.right).k)
                node.right = self.__remove_min(node.right)
                successor.left = node.left
                successor.right = node.right
                return successor

class BST2(object):
    def __init__(self):
        self.root = None
        self.count = 0

    def insert(self, k):
        if not self.root:
            self.root = Node(k)
            return

        node = self.root
        while 1:
            if k == node.k:
                return
            elif k < node.k:
                if node.left:
                    node = node.left
                else:
                    node.left = Node(k)
                    return
            else:
                if node.right:
                    node = node.right
                else:
                    node.right = Node(k)
                    return

    def in_order(self):
        stack = []
        node = self.root
        while node or stack:
            while node:
                stack.append(node)
                node = node.left
            if stack:
                node = stack.pop()
                do_something(node)
                node = node.right

    def pre_order(self):
        stack = []
        node = self.root
        while node or stack:
            while node:
                do_something(node)
                stack.append(node)
                node = node.left
            if stack:
                node = stack.pop()
                node = node.right

    def level_order(self):
        q = []
        node = self.root
        q.append(node)
        while q:
            node = q.pop(0)
            do_something(node)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)

    def contain(self, k):
        node = self.root
        while node:
            if k == node.k:
                return True
            elif k < node.k:
                node = node.left
            else:
                node = node.right
        return False

    def minimum(self):
        return self.__minimum(self.root)

    def __minimum(self,node):
        if not node:
            return
        while node:
            if not node.left:
                node = node.left
            else:
                return node

    def find_father(self,k):
        father = None
        node = self.root
        while node:
            if k == node.k:
                return father
            else:
                father = node
                if k < node.k:
                    node = node.left
                else:
                    node = node.right
        return False



    def remove_min(self):
        self.__remove_min(self.root)

    def __remove_min(self,node):
        if not node:
            return
        while node.left:
            if not node.left.left:
                node.left = node.left.right
                return
            else:
                node = node.left




def test():
    bst = BST()
    bst2 = BST2()
    n = 10
    arr = [randint(0, n - 1) for _ in xrange(n)]
    for i in arr:
        i = randint(0,99)
        bst.insert(i)
        bst2.insert(i)

    # bst.level_order()
    bst2.level_order()
    print '=' * 10

    # bst.in_order()
    # print '=' * 10
    # bst2.in_order()
    # print '*' * 10

    # bst.remove(bst.root.k)
    # bst.level_order()
    print bst2.find_father(bst.minimum().k).k

    # bst2.level_order()


if __name__ == '__main__':
    test()
